### ISSUE TITLE
Std.is cannot detect interfaces on externs

### ISSUE THREAD

**Issue by jcward** [HaxeFoundation/haxe#5947]
Here's the problem:

1) The Haxe type system allows interfaces on externs. This is nice, as it means we can pass externs to methods by interface.
2) Unfortunately, `Std.is` (at least in the JS target, haven't tested others) cannot detect these interfaces. That is, `Std.is(extern_instance, ISomeInterface)` always returns false. While not immediately obvious, that's understandable, as the external class can't be decorated with the metadata required to allow this runtime type checking.

However, these two items, taken together, seem to indicate that the type system makes a promise that the dynamic type tools can't keep... and it isn't documented.

Here's some sample code: http://try.haxe.org/#Ed6b2

Interestingly, `Std.is` (again, at least in JS -- is this cross-platform?) can detect the extern itself (e.g. `Std.is(extern_instance, ExternClass)` *does* return properly) because it uses the [native JavaScript `instanceof` check](url).

I can imagine a couple possible solutions:

- Implement some kind of internal metadata tracking -- e.g. a map of interfaces to externs, such that when testing a given interface, `Std.is` also tests explicitly for the mapped externs.
- Give up, and document the shortcoming in the [`Std.is` documentation](http://api.haxe.org/Std.html#is), e.g. adding a note:

> Note: Std.is cannot detect whether extern instances satisfy an interface.

...

Another interesting effect of this issue, possibly worth documenting, is that unsafe casts from the extern to the interface will succeed, while safe casts will throw. Because safe casts use Std.is to decide when to throw.

**Comment by ibilon** [comment]
If you say it implements an interface haxe will accept it, but that's doesn't add the interface to the extern like you said.
So doesn't it fail because you lie about the extern? The same way it'd fail at runtime/native compilation if you change the type of a function.

**Comment by Simn** [comment]
Agreed, the problem here is that your externs are not accurate. This makes me wonder if extern classes should even be allowed to implement/extend non-extern interfaces/classes.

**Comment by ibilon** [comment]
@jcward You can pass them like they had the interface like in http://try.haxe.org/#7970f
Do you need the run time check?

> Agreed, the problem here is that your externs are not accurate. This makes me wonder if extern classes should even be allowed to implement/extend non-extern interfaces/classes.

Most likely not, doesn't make sense that something native would use a haxe class without having itself been generated from haxe, which in that case you wouldn't need the externs.

**Comment by RealyUniqueName** [comment]
For php target interfaces on externs will be checked correctly. It's just JS (and maybe Lua) does not have such thing like interfaces.
Also in terms of modular Haxe->JS extern actual can implement interfaces.
So, please, don't ban interfaces on externs )

**Comment by Simn** [comment]
Right, there's always the weird "extern for Haxe types" scenario. So I guess there's nothing to be done here and users have to make sure they define their externs properly.

**Comment by RealyUniqueName** [comment]
In that specific scenario provided by @jcward it's more like typedef than interface. 
Maybe Haxe needs something like `Std.unifies(value, Type)` macro which will generate run time checks for each supplied `Type`

**Comment by ncannasse** [comment]
@RealyUniqueName it's not sound do to this. We can't infer for instance if [] is an Array of String or Int or whatever can be pushed later into it

**Comment by jcward** [comment]
In my opinion, I haven't lied about my interface, the native `Date` implements `getTime()` just as I've specified. And interfaces on externs certainly seems like a valuable feature (though potentially with platform-specific concerns.)

So, I don't think this issue deserves to be closed outright without at least considering an update to the documentation.

**Comment by Simn** [comment]
> the native Date implements getTime() just as I've specified

This would only be true in a duck-typing kind of sense, which is not how interface implementation is specified (in Haxe or otherwise).

**Comment by RealyUniqueName** [comment]
You used interface like a typedef. Real interfaces are not about just having the same set of fields, but also about having an explicit runtime reference to implemented type in implementing type.

**Comment by jcward** [comment]
Alright, well, it sounds like I need to resort to magic. Disturbingly, I'm getting quite used to these types of Haxe hacks.

```
@:native("window.Date")
extern class JSDate implements IHasGetTime {
    public function new();
    public function getTime():Float;

    static function __init__() : Void {
        untyped __js__('{0}.__interfaces__ = {1}', JSDate, [ IHasGetTime ]);
    }
}
```

With hack: http://try.haxe.org/#7963C

**Comment by ncannasse** [comment]
@jcward an interface is a named type, not simply a declaration.

For instance the following two interfaces:

```haxe
interface A {
     function foo() : Void;
}
interface B {
     function foo() : Void;
}
```

A is not compatible with B. Classes implementing only A are not of the type B and vice-versa. One class can still implement both A and B, but it requires explicit declaration.

And as a replied earlier, runtime type checking of structures is not sound.

**Comment by jcward** [comment]
> @jcward an interface is a named type, not simply a declaration.

Ah. What a fun bit of strict typing trivia. I'm simply glad there is a simple mechanism by which I can free the Haxe's JS RTTI from this oppressive viewpoint.

**Comment by Simn** [comment]
> oppressive viewpoint

It's not a viewpoint, you simply have a misconception of what interfaces are. This is [documented](https://haxe.org/manual/types-interfaces.html), too:

> Interfaces, unlike structural subtyping, describe a static relation between classes.

**Comment by jcward** [comment]
But more seriously, I accept that what I'm describing doesn't fit into Haxe's type system (that is, an external class, built apart from Haxe, cannot strictly implement a Haxe interface as defined.) So I'll concede that no change is necessary.

However, I am saying there is utility in allowing said external class to exhibit the polymorphic features that interfaces provide. Ignoring the strict definition of interface, and hacking the __interface__ property seems to do exactly that.

This runtime patch of an external class actually helps me avoid duck typing. I'm effectively telling both the compiler and `Std.is` -- trust me, it effectively implements the interface.

**Comment by nadako** [comment]
Actually, since there's no such thing as nominal interface in JavaScript and Haxe emulates it via `__interfaces__` property, I don't think it's *that* hacky to patch external types like that and we could even have an utility function for that somewhere in stdlib's `js` package. However it's not always needed (e.g. I almost never use RTTI-related features and I don't want Haxe to modify anything against my will), so it should be deliberate decision of a concrete user.

**Comment by nadako** [comment]
Oh I missed the "proposed solutions":

> * Implement some kind of internal metadata tracking -- e.g. a map of interfaces to externs, such that when testing a given interface, Std.is also tests explicitly for the mapped externs.

This is an interesting idea to think about, though I'm not sure it's really worth the effort.

> * Give up, and document the shortcoming in the Std.is documentation, e.g. adding a note

That wouldn't hurt.

**Comment by jcward** [comment]
> > some kind of internal metadata tracking

> This is an interesting idea to think about, though I'm not sure it's really worth the effort.

Right, I had that "metadata tracking" idea before the `__interfaces__` workaround.

> > document the shortcoming

> That wouldn't hurt.

Thanks for at least acknowledging that. But I suppose it's not a shortcoming. Though as we've decided, it's kind of a gray area as it's an "impossible" type specification sitting at the boundary between Haxe and JS.

> it should be deliberate decision of a concrete user.

I agree with that. Maybe an `@:metadata`. Or the implementation could only apply the hack if the extern doesn't already have an `__interfaces__` property (external JS compiled by Haxe already would have `__interfaces__` and shouldn't be overwritten.)  e.g.:

```
        untyped __js__('if (typeof {0}.__interfaces__ == "undefined") {0}.__interfaces__ = {1}', JSDate, [ IHasGetTime ]);
```

But again, I found my solution, so I'm not going to lobby further.
